File: src/services/api/maestros.ts
/**
 * Maestros API — 8 catalogs via /maestros/* (Paginated) with tryFetch fallback for canales/metodos.
 * Keeps canales/metodos backward compatible by wrapping static arrays as Paginated when API unavailable.
 */
import { client } from '@/api/client'
import { showToast } from '@/utils/toast'

export interface Paginated<T> {
  items: T[]
  total: number
}

// --- Read shapes (minimal fields used by UI) ---
export interface ProveedorRead {
  id: number
  nombre: string
  categoria: string
  ciudad?: string | null
  calificacion?: number | null
  tiempo_entrega_dias?: number | null
  email?: string | null
  telefono?: string | null
  activo: boolean
  notas?: string | null
  created_at: string
  updated_at: string
}
export interface CategoriaRead {
  id: number
  nombre: string
  tipo_talla: string
  descripcion?: string | null
  margen_meta_pct?: number | null
  total_modelos: number
  activo: boolean
  created_at: string
  updated_at: string
}
export interface UbicacionRead {
  id: number
  codigo: string
  nombre: string
  tipo: string
  capacidad?: string | null
  observaciones?: string | null
  activo: boolean
  created_at: string
  updated_at: string
}
export interface CanalRead {
  id: number
  codigo: string
  nombre: string
  tipo?: string | null
  comision_pct?: number | null
  costo_fijo_mensual?: number | null
  activo?: boolean | null
  descripcion?: string | null
  created_at: string
  updated_at?: string | null
}
export interface MetodoRead {
  id: number
  codigo: string
  nombre: string
  tipo?: string | null
  comision_pct?: number | null
  tiempo_acreditacion?: string | null
  activo?: boolean | null
  datos_cuenta?: string | null
  descripcion?: string | null
  created_at: string
  updated_at?: string | null
}
export interface TallaRead {
  id: number
  talla: string
  orden: number
  busto?: string | null
  cintura?: string | null
  cadera?: string | null
  reduccion_corset?: string | null
  descripcion?: string | null
  activo: boolean
  created_at: string
  updated_at: string
}
export interface ProductoSinTallaRead {
  id: number
  nombre: string
  categoria: string
  dimensiones?: string | null
  materiales?: string | null
  descripcion?: string | null
  precio_sugerido: number
  activo: boolean
  created_at: string
  updated_at: string
}
export interface ParametrosRead {
  id: number
  costo_minuto_costura: number
  costo_hora_patronaje: number
  margen_meta_global_pct: number
  desperdicio_textil_default_pct: number
  iva_regimen_pct: number
  distribucion_reinversion_pct: number
  reparto_margara_pct: number
  reparto_valqui_pct: number
  created_at: string
  updated_at: string
}

// Backwards compat static fallbacks
export interface CanalVentaMaestro { codigo: string; nombre: string }
export interface MetodoPagoMaestro { codigo: string; nombre: string }
export const CANALES_VENTA: CanalVentaMaestro[] = [
  { codigo: 'web', nombre: 'Web' },
  { codigo: 'whatsapp', nombre: 'WhatsApp / DM' },
  { codigo: 'instagram', nombre: 'Instagram' },
  { codigo: 'feria', nombre: 'Feria / Evento' },
  { codigo: 'showroom_pereira', nombre: 'Showroom Pereira' },
]
export const METODOS_PAGO: MetodoPagoMaestro[] = [
  { codigo: 'efectivo', nombre: 'Efectivo' },
  { codigo: 'transferencia', nombre: 'Transferencia' },
  { codigo: 'tarjeta', nombre: 'Tarjeta' },
  { codigo: 'contraentrega', nombre: 'Contraentrega' },
]

// Generic param shape
export interface ListParams {
  q?: string
  limit?: number
  offset?: number
  sort_by?: string
  order?: 'asc' | 'desc'
  [key: string]: unknown
}

async function tryFetch<T>(path: string, fallback: T, params?: Record<string, unknown>): Promise<T> {
  try {
    const { data } = await client.get<T>(path, { params })
    return data
  } catch {
    // P2-2: fail-loud — el fallback estático se mantiene para no romper ventas,
    // pero se avisa en vez de enmascarar la caída del backend.
    console.warn(`[maestros] ${path} no disponible, usando valores locales`)
    try {
      showToast('warn', 'Maestros no disponibles, usando valores locales', path)
    } catch { /* toast sin suscriptores: solo el console.warn vale */ }
    return fallback
  }
}

// Proveedores
export async function listProveedores(params: ListParams = {}): Promise<Paginated<ProveedorRead>> {
  const { data } = await client.get<Paginated<ProveedorRead>>('/maestros/proveedores', { params })
  return data
}
export async function getProveedor(id: number): Promise<ProveedorRead> {
  const { data } = await client.get<ProveedorRead>(`/maestros/proveedores/${id}`)
  return data
}
export async function createProveedor(payload: Record<string, unknown>): Promise<ProveedorRead> {
  const { data } = await client.post<ProveedorRead>('/maestros/proveedores', payload)
  return data
}
export async function updateProveedor(id: number, payload: Record<string, unknown>): Promise<ProveedorRead> {
  const { data } = await client.patch<ProveedorRead>(`/maestros/proveedores/${id}`, payload)
  return data
}
export async function deleteProveedor(id: number): Promise<void> {
  await client.delete(`/maestros/proveedores/${id}`)
}

// Categorias
export async function listCategorias(params: ListParams = {}): Promise<Paginated<CategoriaRead>> {
  const { data } = await client.get<Paginated<CategoriaRead>>('/maestros/categorias-coleccion', { params })
  return data
}
export async function createCategoria(payload: Record<string, unknown>): Promise<CategoriaRead> {
  const { data } = await client.post<CategoriaRead>('/maestros/categorias-coleccion', payload)
  return data
}
export async function updateCategoria(id: number, payload: Record<string, unknown>): Promise<CategoriaRead> {
  const { data } = await client.patch<CategoriaRead>(`/maestros/categorias-coleccion/${id}`, payload)
  return data
}
export async function deleteCategoria(id: number): Promise<void> {
  await client.delete(`/maestros/categorias-coleccion/${id}`)
}

// Ubicaciones
export async function listUbicaciones(params: ListParams = {}): Promise<Paginated<UbicacionRead>> {
  const { data } = await client.get<Paginated<UbicacionRead>>('/maestros/ubicaciones-taller', { params })
  return data
}
export async function createUbicacion(payload: Record<string, unknown>): Promise<UbicacionRead> {
  const { data } = await client.post<UbicacionRead>('/maestros/ubicaciones-taller', payload)
  return data
}
export async function updateUbicacion(id: number, payload: Record<string, unknown>): Promise<UbicacionRead> {
  const { data } = await client.patch<UbicacionRead>(`/maestros/ubicaciones-taller/${id}`, payload)
  return data
}
export async function deleteUbicacion(id: number): Promise<void> {
  await client.delete(`/maestros/ubicaciones-taller/${id}`)
}

// Canales (with fallback)
export async function listCanales(params: ListParams = {}): Promise<Paginated<CanalRead>> {
  const fallback: Paginated<CanalRead> = {
    items: CANALES_VENTA.map((c, i) => ({ id: i + 1, codigo: c.codigo, nombre: c.nombre, created_at: new Date().toISOString(), updated_at: null })) as unknown as CanalRead[],
    total: CANALES_VENTA.length,
  }
  return tryFetch<Paginated<CanalRead>>('/maestros/canales-venta', fallback, params)
}
export async function createCanal(payload: Record<string, unknown>): Promise<CanalRead> {
  const { data } = await client.post<CanalRead>('/maestros/canales-venta', payload)
  return data
}
export async function updateCanal(id: number, payload: Record<string, unknown>): Promise<CanalRead> {
  const { data } = await client.patch<CanalRead>(`/maestros/canales-venta/${id}`, payload)
  return data
}
export async function deleteCanal(id: number): Promise<void> {
  await client.delete(`/maestros/canales-venta/${id}`)
}
export async function listCanalesLegacy(): Promise<CanalVentaMaestro[]> {
  return tryFetch<CanalVentaMaestro[]>('/maestros/canales-venta', CANALES_VENTA)
}

// Metodos (with fallback)
export async function listMetodosPago(params: ListParams = {}): Promise<Paginated<MetodoRead>> {
  const fallback: Paginated<MetodoRead> = {
    items: METODOS_PAGO.map((m, i) => ({ id: i + 1, codigo: m.codigo, nombre: m.nombre, created_at: new Date().toISOString(), updated_at: null })) as unknown as MetodoRead[],
    total: METODOS_PAGO.length,
  }
  return tryFetch<Paginated<MetodoRead>>('/maestros/metodos-pago', fallback, params)
}
export async function createMetodo(payload: Record<string, unknown>): Promise<MetodoRead> {
  const { data } = await client.post<MetodoRead>('/maestros/metodos-pago', payload)
  return data
}
export async function updateMetodo(id: number, payload: Record<string, unknown>): Promise<MetodoRead> {
  const { data } = await client.patch<MetodoRead>(`/maestros/metodos-pago/${id}`, payload)
  return data
}
export async function deleteMetodo(id: number): Promise<void> {
  await client.delete(`/maestros/metodos-pago/${id}`)
}
export async function listMetodosLegacy(): Promise<MetodoPagoMaestro[]> {
  return tryFetch<MetodoPagoMaestro[]>('/maestros/metodos-pago', METODOS_PAGO)
}

// Tallas
export async function listTallas(params: ListParams = {}): Promise<Paginated<TallaRead>> {
  const { data } = await client.get<Paginated<TallaRead>>('/maestros/tallas-estandar', { params })
  return data
}
export async function createTalla(payload: Record<string, unknown>): Promise<TallaRead> {
  const { data } = await client.post<TallaRead>('/maestros/tallas-estandar', payload)
  return data
}
export async function updateTalla(id: number, payload: Record<string, unknown>): Promise<TallaRead> {
  const { data } = await client.patch<TallaRead>(`/maestros/tallas-estandar/${id}`, payload)
  return data
}
export async function deleteTalla(id: number): Promise<void> {
  await client.delete(`/maestros/tallas-estandar/${id}`)
}

// Productos sin talla
export async function listProductosSinTalla(params: ListParams = {}): Promise<Paginated<ProductoSinTallaRead>> {
  const { data } = await client.get<Paginated<ProductoSinTallaRead>>('/maestros/productos-sin-talla', { params })
  return data
}
export async function createProductoSinTalla(payload: Record<string, unknown>): Promise<ProductoSinTallaRead> {
  const { data } = await client.post<ProductoSinTallaRead>('/maestros/productos-sin-talla', payload)
  return data
}
export async function updateProductoSinTalla(id: number, payload: Record<string, unknown>): Promise<ProductoSinTallaRead> {
  const { data } = await client.patch<ProductoSinTallaRead>(`/maestros/productos-sin-talla/${id}`, payload)
  return data
}
export async function deleteProductoSinTalla(id: number): Promise<void> {
  await client.delete(`/maestros/productos-sin-talla/${id}`)
}

// Parametros singleton
export async function getParametros(): Promise<ParametrosRead> {
  const { data } = await client.get<ParametrosRead>('/maestros/parametros-costeo')
  return data
}
export async function updateParametros(payload: Record<string, unknown>): Promise<ParametrosRead> {
  const { data } = await client.patch<ParametrosRead>('/maestros/parametros-costeo', payload)
  return data
}

File: backend/app/schemas/maestros.py
"""Maestros schemas — 7 catalogs + singleton."""
from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field

# Proveedor
class ProveedorCreate(BaseModel):
    nombre: str = Field(max_length=100)
    categoria: str = Field(max_length=100)
    ciudad: str | None = Field(default=None, max_length=80)
    calificacion: Decimal | None = Field(default=None, ge=0, le=5)
    tiempo_entrega_dias: int | None = Field(default=None, ge=0)
    email: EmailStr | None = None
    telefono: str | None = Field(default=None, max_length=50)
    activo: bool | None = True
    notas: str | None = None


class ProveedorUpdate(BaseModel):
    nombre: str | None = Field(default=None, max_length=100)
    categoria: str | None = Field(default=None, max_length=100)
    ciudad: str | None = Field(default=None, max_length=80)
    calificacion: Decimal | None = Field(default=None, ge=0, le=5)
    tiempo_entrega_dias: int | None = Field(default=None, ge=0)
    email: EmailStr | None = None
    telefono: str | None = Field(default=None, max_length=50)
    activo: bool | None = None
    notas: str | None = None


class ProveedorRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    nombre: str
    categoria: str
    ciudad: str | None
    calificacion: Decimal | None
    tiempo_entrega_dias: int | None
    email: str | None
    telefono: str | None
    activo: bool
    notas: str | None
    created_at: datetime
    updated_at: datetime


# Categoria
TipoTalla = Literal["CON_TALLAS_ESTANDAR", "SIN_TALLA_MERCH", "TALLA_UNICA"]


class CategoriaCreate(BaseModel):
    nombre: str = Field(max_length=100)
    tipo_talla: TipoTalla
    descripcion: str | None = None
    margen_meta_pct: Decimal | None = Field(default=None, ge=0, le=100)
    total_modelos: int | None = Field(default=0, ge=0)
    activo: bool | None = True


class CategoriaUpdate(BaseModel):
    nombre: str | None = Field(default=None, max_length=100)
    tipo_talla: TipoTalla | None = None
    descripcion: str | None = None
    margen_meta_pct: Decimal | None = Field(default=None, ge=0, le=100)
    total_modelos: int | None = Field(default=None, ge=0)
    activo: bool | None = None


class CategoriaRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    nombre: str
    tipo_talla: str
    descripcion: str | None
    margen_meta_pct: Decimal | None
    total_modelos: int
    activo: bool
    created_at: datetime
    updated_at: datetime


# Ubicacion
TipoUbicacion = Literal["ROLLOS_TELAS", "GAVETAS_HERRAJES", "PERCHERO_SHOWROOM", "ACCESORIOS_BODEGA"]


class UbicacionCreate(BaseModel):
    codigo: str = Field(max_length=20, pattern=r"^UB-.*")
    nombre: str = Field(max_length=100)
    tipo: TipoUbicacion
    capacidad: str | None = Field(default=None, max_length=100)
    observaciones: str | None = None
    activo: bool | None = True


class UbicacionUpdate(BaseModel):
    codigo: str | None = Field(default=None, max_length=20, pattern=r"^UB-.*")
    nombre: str | None = Field(default=None, max_length=100)
    tipo: TipoUbicacion | None = None
    capacidad: str | None = Field(default=None, max_length=100)
    observaciones: str | None = None
    activo: bool | None = None


class UbicacionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    codigo: str
    nombre: str
    tipo: str
    capacidad: str | None
    observaciones: str | None
    activo: bool
    created_at: datetime
    updated_at: datetime


# Canal
TipoCanal = Literal["FISICO", "DIGITAL", "EVENTO"]


class CanalCreate(BaseModel):
    codigo: str = Field(max_length=50)
    nombre: str = Field(max_length=100)
    tipo: TipoCanal | None = None
    comision_pct: Decimal | None = Field(default=None, ge=0, le=100)
    costo_fijo_mensual: Decimal | None = Field(default=None, ge=0)
    activo: bool | None = True
    descripcion: str | None = None


class CanalUpdate(BaseModel):
    codigo: str | None = Field(default=None, max_length=50)
    nombre: str | None = Field(default=None, max_length=100)
    tipo: TipoCanal | None = None
    comision_pct: Decimal | None = Field(default=None, ge=0, le=100)
    costo_fijo_mensual: Decimal | None = Field(default=None, ge=0)
    activo: bool | None = None
    descripcion: str | None = None


class CanalRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    codigo: str
    nombre: str
    tipo: str | None
    comision_pct: Decimal | None
    costo_fijo_mensual: Decimal | None
    activo: bool | None
    descripcion: str | None
    created_at: datetime
    updated_at: datetime | None


# Metodo Pago
TipoMetodo = Literal["TRANSFERENCIA", "BILLETERA_DIGITAL", "EFECTIVO", "PASARELA_DATAFONO"]


class MetodoCreate(BaseModel):
    codigo: str = Field(max_length=50)
    nombre: str = Field(max_length=100)
    tipo: TipoMetodo | None = None
    comision_pct: Decimal | None = Field(default=None, ge=0, le=100)
    tiempo_acreditacion: str | None = Field(default=None, max_length=50)
    activo: bool | None = True
    datos_cuenta: str | None = None
    descripcion: str | None = None


class MetodoUpdate(BaseModel):
    codigo: str | None = Field(default=None, max_length=50)
    nombre: str | None = Field(default=None, max_length=100)
    tipo: TipoMetodo | None = None
    comision_pct: Decimal | None = Field(default=None, ge=0, le=100)
    tiempo_acreditacion: str | None = Field(default=None, max_length=50)
    activo: bool | None = None
    datos_cuenta: str | None = None
    descripcion: str | None = None


class MetodoRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    codigo: str
    nombre: str
    tipo: str | None
    comision_pct: Decimal | None
    tiempo_acreditacion: str | None
    activo: bool | None
    datos_cuenta: str | None
    descripcion: str | None
    created_at: datetime
    updated_at: datetime | None


# Talla
class TallaCreate(BaseModel):
    talla: str = Field(max_length=20)
    orden: int
    busto: str | None = Field(default=None, max_length=50)
    cintura: str | None = Field(default=None, max_length=50)
    cadera: str | None = Field(default=None, max_length=50)
    reduccion_corset: str | None = Field(default=None, max_length=50)
    descripcion: str | None = None
    activo: bool | None = True


class TallaUpdate(BaseModel):
    talla: str | None = Field(default=None, max_length=20)
    orden: int | None = None
    busto: str | None = Field(default=None, max_length=50)
    cintura: str | None = Field(default=None, max_length=50)
    cadera: str | None = Field(default=None, max_length=50)
    reduccion_corset: str | None = Field(default=None, max_length=50)
    descripcion: str | None = None
    activo: bool | None = None


class TallaRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    talla: str
    orden: int
    busto: str | None
    cintura: str | None
    cadera: str | None
    reduccion_corset: str | None
    descripcion: str | None
    activo: bool
    created_at: datetime
    updated_at: datetime


# Producto sin talla
class ProductoSinTallaCreate(BaseModel):
    nombre: str = Field(max_length=100)
    categoria: str = Field(max_length=100)
    dimensiones: str | None = Field(default=None, max_length=100)
    materiales: str | None = Field(default=None, max_length=200)
    descripcion: str | None = None
    precio_sugerido: Decimal = Field(ge=0)
    activo: bool | None = True


class ProductoSinTallaUpdate(BaseModel):
    nombre: str | None = Field(default=None, max_length=100)
    categoria: str | None = Field(default=None, max_length=100)
    dimensiones: str | None = Field(default=None, max_length=100)
    materiales: str | None = Field(default=None, max_length=200)
    descripcion: str | None = None
    precio_sugerido: Decimal | None = Field(default=None, ge=0)
    activo: bool | None = None


class ProductoSinTallaRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    nombre: str
    categoria: str
    dimensiones: str | None
    materiales: str | None
    descripcion: str | None
    precio_sugerido: Decimal
    activo: bool
    created_at: datetime
    updated_at: datetime


# Parametros singleton
class ParametrosRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    costo_minuto_costura: Decimal
    costo_hora_patronaje: Decimal
    margen_meta_global_pct: Decimal
    desperdicio_textil_default_pct: Decimal
    iva_regimen_pct: Decimal
    distribucion_reinversion_pct: Decimal
    reparto_margara_pct: Decimal
    reparto_valqui_pct: Decimal
    created_at: datetime
    updated_at: datetime


class ParametrosUpdate(BaseModel):
    costo_minuto_costura: Decimal | None = Field(default=None, ge=0)
    costo_hora_patronaje: Decimal | None = Field(default=None, ge=0)
    margen_meta_global_pct: Decimal | None = Field(default=None, ge=0, le=100)
    desperdicio_textil_default_pct: Decimal | None = Field(default=None, ge=0, le=100)
    iva_regimen_pct: Decimal | None = Field(default=None, ge=0, le=100)
    distribucion_reinversion_pct: Decimal | None = Field(default=None, ge=0, le=100)
    reparto_margara_pct: Decimal | None = Field(default=None, ge=0, le=100)
    reparto_valqui_pct: Decimal | None = Field(default=None, ge=0, le=100)
